sdk/cosmos/azure-cosmos/azure/cosmos/_execution_context/aio/non_streaming_order_by_aggregator.py

```diff
@@ -0,0 +1,152 @@
+            if isinstance(self._query, dict):
+                # this is a parameterized query, collect all the parameters
+                query = dict(self._query)
+                query["query"] = rewritten_query
+            else:
+                query = rewritten_query
+        else:
+            query = self._query
+
+        return document_producer._DocumentProducer(
+            partition_key_target_range,
+            self._client,
+            self._resource_link,
+            query,
+            self._document_producer_comparator,
+            self._options,
+        )
+
+    async def _get_target_partition_key_range(self):
+
+        query_ranges = self._partitioned_query_ex_info.get_query_ranges()
+        return await self._routing_provider.get_overlapping_ranges(
+            self._resource_link, [routing_range.Range.ParseFromDict(range_as_dict) for range_as_dict in query_ranges]
+        )
+
+    async def _configure_partition_ranges(self):
+        # will be a list of (partition_min, partition_max) tuples
+        targetPartitionRanges = await self._get_target_partition_key_range()
+
+        targetPartitionQueryExecutionContextList = []
+        for partitionTargetRange in targetPartitionRanges:
+            # create and add the child execution context for the target range
+            targetPartitionQueryExecutionContextList.append(
+                self._createTargetPartitionQueryExecutionContext(partitionTargetRange)
+            )
+
+        self._doc_producers = []
+        for targetQueryExContext in targetPartitionQueryExecutionContextList:
+            try:
+                await targetQueryExContext.peek()
+                self._doc_producers.append(targetQueryExContext)
+            except exceptions.CosmosHttpResponseError as e:
+                if exceptions._partition_range_is_gone(e):
+                    # repairing document producer context on partition split
+                    await self._repair_document_producer()
+                else:
+                    raise
+
+            except StopAsyncIteration:
+                continue
+
+        pq_size = self._partitioned_query_ex_info.get_top() or self._partitioned_query_ex_info.get_limit()
+        for doc_producer in self._doc_producers:
+            while True:
+                try:
+                    result = await doc_producer.peek()
+                    item_result = document_producer._NonStreamingItemResultProducer(result, self._sort_orders)
+                    await self._orderByPQ.push_async(item_result, self._document_producer_comparator)
+                    await doc_producer.__anext__()
+                except StopAsyncIteration:
+                    # this logic is necessary so that we only hold 2 * items_per_partition in memory at any time
+                    if len(self._orderByPQ._heap) > pq_size:
+                        new_heap = []
+                        for i in range(pq_size):  # pylint: disable=unused-variable
```

Comment:
curious: why not do this checking during the push_async stage

Reply:
the underlying list serving as the heap is not fully ordered in its normal state - as such, if I were to add logic like this to `push_async`, we would have to re-make the entire heap doing pop the way we do here to get the proper ordering